src/utxo/db.rs

```diff
@@ -0,0 +1,132 @@
+use crate::bitcoin::consensus::serialize;
+use crate::bitcoin::{Block, OutPoint, TxOut};
+use crate::utxo::{Hash32, Hash64, Utxo};
+use bitcoin::consensus::deserialize;
+use log::debug;
+use rocksdb::{DBCompressionType, Options, WriteBatch, DB};
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::path::Path;
+
+pub struct DbUtxo {
+    db: DB,
+    updated_up_to_height: i32,
+    inserted_outputs: u64,
+}
+
+impl DbUtxo {
+    pub fn new<P: AsRef<Path>>(path: P) -> Result<DbUtxo, rocksdb::Error> {
+        let mut options = Options::default();
+        options.set_compression_type(DBCompressionType::Snappy);
+        options.create_if_missing(true);
+        let db = DB::open(&options, path)?;
+
+        let updated_up_to_height = db
+            .get("updated_up_to_height")?
+            .map(|e| e.try_into().unwrap())
+            .map(|e| i32::from_ne_bytes(e))
+            .unwrap_or(-1);
+
+        Ok(DbUtxo {
+            db,
+            updated_up_to_height,
+            inserted_outputs: 0,
+        })
+    }
+}
+
+impl Utxo for DbUtxo {
+    fn add(&mut self, block: &Block, height: u32) {
+        let height = height as i32;
+        debug!(
+            "height: {} updated_up_to: {}",
+            height, self.updated_up_to_height
+        );
+        if height > self.updated_up_to_height {
+            // since we can spend outputs created in this same block, we first put outputs in memory...
+            let total_outputs = block.txdata.iter().map(|e| e.output.len()).sum();
+            let mut current_outputs = HashMap::with_capacity(total_outputs);
+            for tx in block.txdata.iter() {
+                let txid = tx.txid();
+                for (i, output) in tx.output.iter().enumerate() {
+                    let outpoint = OutPoint::new(txid, i as u32);
+                    current_outputs.insert(outpoint, output);
+                }
+            }
+
+            let total_inputs = block.txdata.iter().skip(1).map(|e| e.input.len()).sum();
+            let mut prevouts = Vec::with_capacity(total_inputs);
+            let mut batch = WriteBatch::default();
+            for tx in block.txdata.iter().skip(1) {
+                for input in tx.input.iter() {
+                    //...then we first check if inputs spend output created in this block
+                    match current_outputs.remove(&input.previous_output) {
+                        Some(tx_out) => {
+                            // we avoid touching the db entirely if it's spent in the same block
+                            prevouts.push(tx_out.clone())
+                        }
+                        None => {
+                            let key = input.previous_output.to_key();
+                            let tx_out = deserialize(&self.db.get(&key).unwrap().unwrap()).unwrap();
```

Comment:
I would considering slapping LRU somewhere in `DbUtxo` if `self.db` doesn't already do some great caching. Seems to me that a lot of outputs are spent maybe not in the same block, but very soon after, so a small cache could give a nice perf boost. Though maybe rocksdb already does some caching which renders this not very important. It's much useful when using remote db, I guess.